python_modules/dagster/dagster/_core/execution/plan/compute_generator.py

```diff
@@ -133,14 +134,32 @@ def _coerce_op_compute_fn_to_iterator(
 def _zip_and_iterate_op_result(
     result: Any, context: OpExecutionContext, output_defs: Sequence[OutputDefinition]
 ) -> Iterator[Tuple[int, Any, OutputDefinition]]:
-    if len(output_defs) > 1:
-        result = _validate_multi_return(context, result, output_defs)
-        for position, (output_def, element) in enumerate(zip(output_defs, result)):
+    expected_return_outputs = _filter_expected_output_defs(result, context, output_defs)
+    if len(expected_return_outputs) > 1:
+        result = _validate_multi_return(context, result, expected_return_outputs)
+        for position, (output_def, element) in enumerate(zip(expected_return_outputs, result)):
             yield position, output_def, element
     else:
         yield 0, output_defs[0], result
 
 
+# Filter out output_defs corresponding to asset check results that already exist on a
+# MaterializeResult.
+def _filter_expected_output_defs(
+    result: Any, context: OpExecutionContext, output_defs: Sequence[OutputDefinition]
+) -> Sequence[OutputDefinition]:
+    result_tuple = (
+        (result,) if not isinstance(result, tuple) or is_named_tuple_instance(result) else result
+    )
+    materialize_results = [x for x in result_tuple if isinstance(x, MaterializeResult)]
+    remove_outputs = [
+        r.get_spec_python_identifier(x.asset_key or context.asset_key)
+        for x in materialize_results
+        for r in x.check_results or []
+    ]
+    return [out for out in output_defs if out.name not in remove_outputs]
+
+
```

Comment:
I just thought on alternative approach here. What if we unpacked `MaterializeResult` right in `_zip_and_iterate_op_result` and yielded the appropriate `AssetCheckResult` and `Output` objects right there.

Reply:
Maybe doable but not trivial because the logic that currently converts `MaterializeResult` to `Output` in `execute_step.py` accesses `job_def.asset_layer`, which is not available for direct invocation, and `_zip_and_iterate_op_result` is hit in direct invocation.

Recommend exploring this in the proposed followup refactor rather than this PR.

Reply:
Ok. Please leave comment in `_zip_and_iterate_op_result` that contains all the context in your head about this problematic code path (and is appropriately apologetic in tone :-))

This is a fairly risk change and I want any future people who have to deal with this before you fix it to understand all that context.